### DISCUSSION TITLE
ThisAssembly.Resources

### DISCUSSION THREAD

**Discussion by viceroypenguin** [devlooped/ThisAssembly#34] [Ideas]
Proposal: Add another project for exposing `string` and `Stream` versions of embedded resources. Accessing the resources is something necessary for most source generators, but also for many other projects; you're accessing embedded resources already in the other ThisAssembly.* projects. It would be nice to be able to have compile-time named access to the embedded resources. Thoughts?

**Comment by kzu** [comment]
I really like this idea. I guess the API cannot be property-based since you'd need to dispose the embedded resource stream after using them, right?

**Reply by viceroypenguin** [reply]
I don't see any reason it couldn't be property-based? Since the stream itself is `IDisposable`, then no reason the stream can't be returned as a new creation every time the property is accessed and let the user dispose when they want to. I don't think there's any concurrency/multi-use restrictions on the embedded resource streams as long as a new one is created every time.

[removed]

**Reply by viceroypenguin** [reply]
Sorry, that was incorrect, I think. Since it is a stream, it should be method based yes. This is probably more accurate code:

```csharp
class Resources
{
	private static string _namedResourceText;
	public static string NamedResourceText
	{
		get
		{
			return _namedResourceText ??=
				_();
			static string _()
			{
				using (var str = GetNamedResourceStream())
				using (var sr = new StreamReader(str))
					return str.ReadToEnd();
			}
		}
	}

	public static Stream GetNamedResourceStream()
	{
		// already written elsewhere
	}
}
```

**Reply by kzu** [reply]
I wonder if it might be better to have: `byte[] ThisAssembly.Resources.MyResource { get; }` (cached as shown above) and then a separate `Stream ThisAssembly.Resources.GetStream(ResourceName name)` where the `ResourceName` is a generated enum...  

Since it's nearly impossible to know if a given embedded resource is a string or not, generating a text member for all of them seems inapropriate. Also, users could still manually create and expose a text-based one easily, like so:

```
partial class Resources
{
   static string? myResourceText;
   public static string MyResourceText => myResourceText ??= Encoding.UTF8.GetString(ThisAssembly.Resources.MyResource);
}
```

So that would show alongside ` ThisAssembly.Resources.MyResource` in intellisense....

**Reply by viceroypenguin** [reply]
I think there are probably enough text files embedded that are desired as text, that it would be nice to have some level of shortcut to remove the boilerplate. 

What about this:
```csharp
public class Resources
{
  public class MyResource
  {
    byte[] Bytes { get; }
    string Text { get; }
    Stream GetStream();
  }
}
```

Adds another `.`, but would simplify and make explicit the decision of which format to get the data in?

**Reply by viceroypenguin** [reply]
Alternatively, could we use a custom flag in the `<Resources />` tag in the .csproj to identify whether or not it should provide a `string` version?

**Reply by kzu** [reply]
Yeah, that might be better. It's a tricky thing because those embedded resources can be large, and keeping them cached in those properties might adversely affect the app's performance (thinking mobile, for example). An explicit opt-in via metadata sounds reasonable. Perhaps it could even expose an additional `Cached` attribute to generate a property that fetches the resource every time VS caches the result?

**Reply by viceroypenguin** [reply]
I'm good w/ a `Cached` attribute defaulting to `true` as well. That definitely makes sense for situations w/ low-access high-size resources.

**Reply by kzu** [reply]
Hm... maybe you meant the opposite? w/low-access, high-size risk, defaulting to `false` might make more sense?

So, how about:

```xml
<EmbeddedResource Include="..." Type="String|Bytes|Bitmap|Icon|...others from System.Drawing.Common?" Cached="true|false" />
```

I'd make cached *false* by default. It can be changed easily enough with a bit of MSBuild for all your resources if they are all small, like:

```xml
<ItemDefinitionGroup>
   <EmbeddedResource>
     <Cached>true</Cached>
   <EmbeddedResource>
</ItemDefinitionGroup>
```

Or you can update individual elements by extension, like:

```xml
<ItemGroup>
   <EmbeddedResource Update="@(EmbeddedResource -> WithMetadataValue('Extension', '.txt'))">
     <Cached>true</Cached>
   <EmbeddedResource>
</ItemGroup>
```

and so on.

**Reply by viceroypenguin** [reply]
Sorry, I mixed up two different opinions. 

I agree w/ adding a `Cached` attribute because of low-access high-size situations. 

That said, I think a majority of situations are low-risk and would want to be cached. I think non-cached would be the exception, so default to `true` to cover the most common case.

However, I'm not particularly beholden to that opinion, so I'm good w/ a default of `false` if you think the rare case is risky enough to not want to default to `true`. I would just want to make sure we have it documented in the readme so users know the library can cache it for them if they wave the magic wand and speak the magic incantation.

I'd say we always expose a `Stream GetStream()` and a `byte[] Bytes {get;set;}`. Since resources are always byte streams regardless of whatever else they mean, these two are always valid interpretations. Then we add `Type="String"` as an option, and maybe the only one for now. Then add others later as users express interest?

**Reply by kzu** [reply]
Good points. I'd actually make it `byte[] GetBytes()` in addition to `GetStream()` and never cache those. Then add `Type="string"` and that means cache by default, unless `Cached=false`. For strings, I think caching makes sense. For arbitrary bytes, less so, so I'd keep that as a method (kinda implies it's not cached). 

And maybe we could have a default set of extensions that are considered `Type=string`, like `.txt;.cs;.scbn;.log` and the like? That would likely be another property you can set, something like `EmbeddedResourceStringExtensions=...` you can append to...

**Reply by viceroypenguin** [reply]
I'm good with that. Do you want any help making this happen?

**Reply by kzu** [reply]
Help is always very much appreciated! I'd take a PR in a pinch 👍

**Reply by viceroypenguin** [reply]
I'll see if I can make some time next weekend to figure out how you're implementing things (as opposed to how I've been implementing them... 😊) and draw something up. Got some other projects I need to get through first though. If you get there before I do, no worries.

**Reply by kzu** [reply]
Ok, summarized the feature at https://github.com/devlooped/ThisAssembly/issues/45. Comments welcomed!

**Comment by rdeago** [comment]
It's a pity this topic (apparently) lost traction.

@kzu are you still in for a PR? I'm currently standardizing on using `ThisAssembly` in all my projects and `Resources` would be a very welcomed enhancement, so much so that I'm willing to invest the time to implement it myself.

**Reply by kzu** [reply]
Yes, I'd absolutely like to have this built in 💪

**Reply by kzu** [reply]
This has now shipped!

**Reply by viceroypenguin** [reply]
Thanks!! :)
